tests/test_0390-virtual-forms.py

```diff
@@ -0,0 +1,108 @@
+    y = x1 * 10.
+    yv = awkward1.virtual(lambda: gen(y, "y"), length=len(y), form=y.layout.form)
+    array = awkward1.zip({"x": xv, "y": yv}, with_name="Point", depth_limit=1)
+    assert array.layout.purelist_parameter("__record__") == "Point"
+    virtual = awkward1.virtual(lambda: gen(array, "array"), length=len(array), form=array.layout.form)
+    assert virtual.layout.purelist_parameter("__record__") == "Point"
+    assert len(materialize_count) == 0
+
+    slicedvirtual = virtual.x
+    assert len(materialize_count) == 0
+    assert slicedvirtual.layout.purelist_parameter("__record__") == "xthing"
+    assert len(materialize_count) == 0
+
+    slicedvirtual = virtual[["x", "y"]]
+    assert len(materialize_count) == 0
+    assert slicedvirtual.layout.purelist_parameter("__record__") == None
+    assert len(materialize_count) == 0
+
+    slicedvirtual = virtual[::2]
+    assert len(materialize_count) == 0
+    assert slicedvirtual.layout.purelist_parameter("__record__") == "Point"
+    assert len(materialize_count) == 0
+
+    slicedvirtual = virtual[:3]
+    assert len(materialize_count) == 0
+    assert slicedvirtual.layout.purelist_parameter("__record__") == "Point"
+    assert len(materialize_count) == 0
+
+    slicedvirtual = virtual[awkward1.Array([True, False, False, True, False])]
+    assert len(materialize_count) == 0
+    assert slicedvirtual.layout.purelist_parameter("__record__") == "Point"
+    assert len(materialize_count) == 0
+
+
+def test_virtual_slice_numba():
+    numba = pytest.importorskip("numba")
+    materialize_count = defaultdict(int)
+
+    def gen(x, name):
+        materialize_count[name] += 1
+        return x
+
+    x1 = awkward1.Array([1, 2, 3, 4, 5])
+    x2 = awkward1.Array([1, 2, 3, 4, 5])
+    x = awkward1.zip({"x1": x1, "x2": x2}, with_name="xthing")
+    xv = awkward1.virtual(lambda: gen(x, "x"), length=len(x), form=x.layout.form)
+    y = x1 * 10.
+    yv = awkward1.virtual(lambda: gen(y, "y"), length=len(y), form=y.layout.form)
+    array = awkward1.zip({"x": xv, "y": yv}, with_name="Point", depth_limit=1)
+    virtual = awkward1.virtual(lambda: gen(array, "array"), length=len(array), form=array.layout.form)
+
+
+    @numba.njit
+    def dostuff(array):
+        x = 0
+        for item in array:
+            x += item
+        return x
+
+    assert dostuff(virtual.x.x1) == 15
+    assert dict(materialize_count) == {"x": 3, "array": 3}
+    materialize_count.clear()
+
+    @numba.njit
+    def dostuff(array):
+        x = 0
+        for item in array:
+            x += item.x.x1
+        return x
+
+    assert dostuff(virtual[["x"]]) == 15
+    assert dict(materialize_count) == {"x": 1, "array": 2}
+    materialize_count.clear()
+
+    assert dostuff(virtual[::2]) == 9
+    assert dict(materialize_count) == {"x": 1, "array": 2}
+    materialize_count.clear()
+
+    assert dostuff(virtual[:3]) == 6
+    assert dict(materialize_count) == {"x": 1, "array": 1}
```

Comment:
I can't say I understand all these materialization passes and why they differ, but since a cache would obviate any concerns about them I'm inclined to leave it be. Someday an update might change these numbers and make these tests fail unnecessarily, but maybe it will be good to draw attention anyway.

Reply:
I agree.

It could be that the number of materializations is multiplied by Numba's type-checking passes. Numba runs our "typer" functions a surprising number of times. At least 3 for each type (part of which is checking const and non-const types...).

But anyway, that's why we have a cache.